internal/pkg/task/env_runner.go

```diff
@@ -0,0 +1,137 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package task
+
+import (
+	"fmt"
+	"github.com/aws/copilot-cli/internal/pkg/aws/ec2"
+	"github.com/aws/copilot-cli/internal/pkg/aws/ecs"
+	"github.com/aws/copilot-cli/internal/pkg/deploy"
+)
+
+const (
+	clusterResourceType  = "ecs:cluster"
+
+	fmtErrClusterFromEnv        = "get cluster by env %s: %w"
+	fmtErrNoClusterFoundFromEnv = "no cluster found in env %s"
+	fmtErrMoreThanOneClusterFromEnv = "more than one cluster is found in environment %s"
+	fmtErrPublicSubnetsFromEnv  = "get public subnet IDs from environment %s: %w"
+	fmtErrSecurityGroupsFromEnv = "get security groups from environment %s: %w"
+)
+
+// Names for tag filters
+var (
+	tagFilterNameForApp = fmt.Sprintf(ec2.TagFilterName, deploy.AppTagKey)
+	tagFilterNameForEnv = fmt.Sprintf(ec2.TagFilterName, deploy.EnvTagKey)
+)
+
+// EnvRunner can run an Amazon ECS task in the VPC and the cluster of an environment.
+type EnvRunner struct {
+	// Count of the tasks to be launched.
+	Count int
+	// Group Name of the tasks that use the same task definition.
+	GroupName string
+
+	// App and Env in which the tasks will be launched.
+	App string
+	Env string
+
+	// Interfaces to interact with dependencies. Must not be nil.
+	VPCGetter     VPCGetter
+	ClusterGetter ResourceGetter
+	Starter       TaskRunner
+}
+
+// Run runs tasks in the environment of the application, and returns the task ARNs.
+func (r *EnvRunner) Run() ([]string, error) {
+	if err := r.validateDependencies(); err != nil {
+		return nil, err
+	}
+
+	cluster, err := r.cluster(r.App, r.Env)
+	if err != nil {
+		return nil, err
+	}
+
+	filters := r.filtersForVPCFromAppEnv()
+
+	subnets, err := r.VPCGetter.PublicSubnetIDs(filters...)
+	if err != nil {
+		return nil, fmt.Errorf(fmtErrPublicSubnetsFromEnv, r.Env, err)
+	}
+	if len(subnets) == 0 {
+		return nil, errNoSubnetFound
+	}
+
+	securityGroups, err := r.VPCGetter.SecurityGroups(filters...)
+	if err != nil {
+		return nil, fmt.Errorf(fmtErrSecurityGroupsFromEnv, r.Env, err)
+	}
+
+	arns, err := r.Starter.RunTask(ecs.RunTaskInput{
+		Cluster:        cluster,
+		Count:          r.Count,
+		Subnets:        subnets,
+		SecurityGroups: securityGroups,
+		TaskFamilyName: taskFamilyName(r.GroupName),
+		StartedBy:      startedBy,
+	})
+	if err != nil {
+		return nil, &errRunTask{
+			groupName: r.GroupName,
+			parentErr: err,
+		}
+	}
+	return arns, nil
+}
+
+func (r *EnvRunner) cluster(app, env string) (string, error) {
+	clusters, err := r.ClusterGetter.GetResourcesByTags(clusterResourceType, map[string]string{
+		deploy.AppTagKey: app,
+		deploy.EnvTagKey: env,
+	})
+
+	if err != nil {
+		return "", fmt.Errorf(fmtErrClusterFromEnv, env, err)
+	}
+
+	if len(clusters) == 0 {
+		return "", fmt.Errorf(fmtErrNoClusterFoundFromEnv, env)
+	}
+
+	// NOTE: only one cluster is associated with an application and an environment
+	if len(clusters) > 1 {
```

Comment:
It may be worth adding a comment here about the possible behaviors we could exhibit if there is in fact more than one cluster. Since we don't currently deploy more than one cluster, we can error out here and it will be fine. But if we deploy more than one, it might be worth checking earlier eventually and adding a SelectOne prompt to pick the cluster in which to run the task.

Reply:
I like what @Louise15926 has done, returning an error if there's more than one. I'm hesitant to say we should add a selector for something that can be a potential future behaviour, since we don't know if this will ever be an issue. Might be something for the future, but don't think we should worry about that at the moment.

Reply:
Yes in the future if we deploy more than one cluster, then this should be checked earlier instead of doing it in `Run`. Agreed!